src/Processors/IProcessor.h

```diff
@@ -299,14 +300,33 @@ class IProcessor
     IQueryPlanStep * getQueryPlanStep() const { return query_plan_step; }
     size_t getQueryPlanStepGroup() const { return query_plan_step_group; }
 
+    uint64_t getElapsedUs() const { return elapsed_us; }
+    uint64_t getNeedDataElapsedUs() const { return need_data_elapsed_us; }
+    uint64_t getPortFullElapsedUs() const { return port_full_elapsed_us; }
+
 protected:
     virtual void onCancel() {}
 
 private:
+    /// For:
+    /// - elapsed_us
+    friend class ExecutionThreadContext;
+    /// For
+    /// - need_data_elapsed_us
+    /// - port_full_elapsed_us
+    friend class ExecutingGraph;
```

Comment:
Note, that right now, I decided to make measurements in the callers instead of in IProcessor itself, this is to avoid lots of changes in the interface, to make review easier.
But I'm not against of change the IProcess interface to encapsulate the logic in it.

Reply:
Yes, I would definitely not touch the processors's interface.

Reply:
I would actually do not add anything new to IProcessor

Reply:
Extracting this information may be too complex, I think putting it into processors is OK, plus it will allow to see these metrics in debugger too.
Objections?